# In This Notebook

Exploratory data analysis of beer object from brewery-db API.

# Setup

In [2]:
import os

from bkcharts import BoxPlot, Histogram, output_notebook, show
from bokeh.models import Range1d
import numpy as np
import pandas as pd

In [3]:
output_notebook()

Loading BokehJS ...

In [4]:
wrk = '../../../data/wrk/brewery-db/'

# Load Data

In [5]:
# read csv data into dataframe object
fpath = os.path.abspath(os.path.join(wrk, 'beers.csv'))
beers = pd.read_csv(fpath)

In [6]:
beers.head(1000)

abv beerVariationId           createDate  \
0    11.10             NaN  2013-01-20 21:38:15   
1      NaN             NaN  2012-01-03 02:42:36   
2      NaN             NaN  2014-05-06 13:08:12   
3      NaN             NaN  2012-01-03 02:42:37   
4    11.50             NaN  2012-01-03 02:42:36   
5     6.30             NaN  2015-04-16 19:04:45   
6     3.50             NaN  2012-01-03 02:42:37   
7    11.40             NaN  2012-01-03 02:42:36   
8     5.00             NaN  2012-01-03 02:42:36   
9     6.90             NaN  2014-06-03 11:51:34   
10     NaN             NaN  2013-03-19 02:18:43   
11    5.30             NaN  2017-06-14 13:31:29   
12    9.50             NaN  2017-04-28 20:00:27   
13    7.50             NaN  2015-08-28 14:09:33   
14   10.00             NaN  2013-02-12 11:44:26   
15   11.64             NaN  2014-09-10 16:07:02   
16    8.50             NaN  2015-11-03 17:02:33   
17   10.00             NaN  2015-06-12 15:16:39   
18    5.10             NaN  2012-01-03 02:42:36   
19   10.10             NaN  2013-02-12 11:45:03   
20    5.80             NaN  2016-05-11 15:05:00   
21    6.00             NaN  2012-01-03 02:42:36   
22    4.80             NaN  2014-04-09 22:04:00   
23    8.50             NaN  2012-09-06 12:44:39   
24    7.60             NaN  2013-07-27 14:04:42   
25   10.00             NaN  2015-07-17 20:19:44   
26    7.00             NaN  2015-07-21 17:27:18   
27    5.70             NaN  2015-07-28 22:37:57   
28    5.80             NaN  2012-03-21 02:01:47   
29    9.50             NaN  2012-01-09 18:45:11   
..     ...             ...                  ...   
970    NaN             NaN  2015-04-09 14:51:31   
971   4.90             NaN  2012-08-12 12:11:38   
972   4.90             NaN  2015-05-12 16:58:55   
973   8.00             NaN  2016-12-08 19:41:42   
974    NaN             NaN  2015-07-13 19:58:29   
975   6.50             NaN  2013-07-23 12:00:16   
976   6.10             NaN  2015-04-24 17:39:42   
977   5.30             NaN  2016-10-23 19:14:53   
978    NaN             NaN  2012-01-03 02:42:42   
979   8.50             NaN  2015-06-11 15:12:24   
980   6.50             NaN  2015-05-15 17:31:19   
981   3.90             NaN  2016-08-15 13:11:10   
982  13.50             NaN  2017-03-29 20:35:38   
983   7.20             NaN  2015-11-09 15:56:27   
984   5.70             NaN  2012-01-03 02:42:42   
985    NaN             NaN  2017-05-22 15:21:18   
986   6.50             NaN  2015-02-16 20:33:49   
987   6.60             NaN  2013-05-29 17:39:19   
988   6.00             NaN  2012-01-03 02:42:42   
989    NaN             NaN  2012-01-03 02:42:42   
990   8.95             NaN  2014-09-19 16:11:34   
991    NaN             NaN  2015-07-21 21:20:09   
992   5.10             NaN  2015-05-18 15:24:37   
993   6.50             NaN  2015-09-24 16:57:31   
994    NaN             NaN  2013-03-01 11:15:34   
995    NaN             NaN  2013-06-27 15:37:20   
996   4.30             NaN  2016-07-11 11:31:27   
997   9.00             NaN  2012-01-03 02:42:42   
998   6.00             NaN  2012-01-03 02:42:42   
999   6.50             NaN  2013-11-17 01:07:07   

                                           description          foodPairings  \
0    Hop Heads this one's for you!  Checking in wit...                   NaN   
1                                                  NaN                   NaN   
2                                                  NaN                   NaN   
3                                                  NaN                   NaN   
4                                                  NaN                   NaN   
5    Stephen Fraley was discussing the Civil War wh...                   NaN   
6                                                  NaN                   NaN   
7                 Der Rauch Gott (Imperial Smoked Ale)                   NaN   
8    A refreshing, golden-hued European-style pilsn...                   NaN   
9    Man Eating Sharks Infest the Atlantic!\

# Describe Data

In [7]:
beers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16091 entries, 0 to 16090
Data columns (total 23 columns):
abv                          13263 non-null float64
beerVariationId              97 non-null object
createDate                   16091 non-null object
description                  9747 non-null object
foodPairings                 809 non-null object
glasswareId                  4137 non-null float64
ibu                          5902 non-null float64
id                           16091 non-null object
ingredients                  440 non-null object
isOrganic                    16091 non-null object
labels                       5551 non-null object
name                         16091 non-null object
nameDisplay                  16091 non-null object
originalGravity              1192 non-null float64
servingTemperature           930 non-null object
servingTemperatureDisplay    930 non-null object
socialAccounts               3867 non-null object
srmId                        1577 non

In [8]:
beers.describe()

abv  glasswareId          ibu  originalGravity        srmId  \
count  13263.000000  4137.000000  5902.000000      1192.000000  1577.000000   
mean       6.600605     5.339618    41.110826         1.061978    17.253012   
std        3.423505     1.489635    26.827465         0.022852    13.531315   
min        0.010000     1.000000     1.000000         1.000000     1.000000   
25%        5.100000     5.000000    22.000000         1.050000     6.000000   
50%        6.000000     5.000000    33.450000         1.058000    12.000000   
75%        7.500000     5.000000    55.450000         1.070000    28.000000   
max      308.000000    14.000000   430.000000         1.452000    41.000000   

            styleId         year  
count  15396.000000   178.000000  
mean      58.959795  2012.185393  
std       40.031253     4.028768  
min        1.000000  1994.000000  
25%       30.000000  2011.000000  
50%       43.000000  2013.000000  
75%       83.000000  2015.000000  
max      170.000000  2017.000000

# Transform Data

In [9]:
def transform_row(row):
    row.createDate = pd.to_datetime(row.createDate)
    if row.isOrganic == 'N':
        row.isOrganic = 0
    elif row.isOrganic == 'Y':
        row.isOrganic = 1
    return row

beers = beers.apply(lambda x: transform_row(x), axis=1)
beers['createYear'] = beers['createDate'].dt.year

# Exploratory Data Analysis

## ABV

The ABV for the beers object appears to have a negatively skewed heteroskedastic distribution, with median of 6.0 ABV and a mean of 6.60 ABV. This suggests there are a relatively small number of high-ABV beers within the distribution. In fact, there are 2 beers, in particular, with greater than 40.0 ABV. There are also 6,344 beers with no description. These might also need to be removed.

In [18]:
beers.abv.describe(percentiles=[.025,.25,.5,.75,.975])

count    13263.000000
mean         6.600605
std          3.423505
min          0.010000
2.5%         4.000000
25%          5.100000
50%          6.000000
75%          7.500000
97.5%       11.500000
max        308.000000
Name: abv, dtype: float64

In [19]:
hist = Histogram(beers[pd.notnull(beers.abv)].abv, title = 'ABV Distribution', bins=100,
              x_range=Range1d(0,50))
show(hist)

In [20]:
high_abv = beers[(pd.notnull(beers.abv)) & (beers.abv > 11.4)]
high_abv.sort_values('abv')

abv beerVariationId          createDate  \
4       11.50             NaN 2012-01-03 02:42:36   
10576   11.50             NaN 2012-01-03 02:43:50   
10537   11.50             NaN 2012-01-03 02:43:50   
10382   11.50             NaN 2014-09-10 00:59:15   
1534    11.50             NaN 2016-05-24 13:20:34   
1616    11.50             NaN 2015-09-01 03:27:09   
8901    11.50             NaN 2012-03-20 02:10:23   
8786    11.50             NaN 2017-01-03 21:27:15   
8193    11.50             NaN 2013-11-07 05:43:49   
7757    11.50             NaN 2017-06-08 13:32:01   
7750    11.50             NaN 2013-11-02 15:46:11   
7748    11.50             NaN 2016-04-03 01:43:02   
2334    11.50             NaN 2012-01-03 02:42:50   
6936    11.50             NaN 2015-04-03 16:18:12   
6869    11.50             NaN 2012-01-03 02:43:20   
2406    11.50             NaN 2012-01-03 02:42:51   
6860    11.50             NaN 2014-07-31 17:50:27   
6856    11.50             NaN 2014-07-31 17:48:59   
6846    11.50             NaN 2014-07-31 17:46:17   
6843    11.50             NaN 2014-07-30 21:02:33   
6841    11.50             NaN 2014-07-30 21:00:18   
6819    11.50             NaN 2012-01-03 02:43:20   
6817    11.50             NaN 2012-01-03 02:43:20   
2718    11.50             NaN 2017-03-26 01:10:31   
5679    11.50             NaN 2015-12-07 19:59:56   
5519    11.50             NaN 2014-01-07 05:18:45   
2901    11.50             NaN 2015-04-03 17:26:29   
2914    11.50             NaN 2016-05-11 00:24:54   
2954    11.50          ySJ2PP 2014-02-18 12:21:38   
4609    11.50             NaN 2017-02-20 16:00:06   
...       ...             ...                 ...   
2539    15.90             NaN 2014-10-20 14:22:05   
12104   16.00             NaN 2015-03-22 20:27:33   
6899    16.00             NaN 2015-08-14 14:53:27   
7564    16.00             NaN 2014-11-28 19:27:25   
15114   16.00             NaN 2012-01-03 02:44:27   
8429    16.00             NaN 2014-07-08 18:52:03   
7327    16.30             NaN 2014-08-27 12:20:47   
3850    16.40             NaN 2015-12-24 23:40:46   
2534    16.50             NaN 2014-10-20 14:27:49   
3810    16.70             NaN 2016-08-03 21:28:07   
13755   16.90             NaN 2014-03-25 13:24:44   
7210    17.00             NaN 2015-11-05 20:28:18   
2898    17.20             NaN 2013-11-07 05:53:56   
14942   17.42             NaN 2012-09-14 13:24:44   
14826   17.50             NaN 2014-08-26 17:44:39   
9838    18.00             NaN 2015-02-23 20:49:59   
87      18.00             NaN 2012-01-03 02:42:36   
10650   18.50             NaN 2016-07-24 00:18:31   
2108    18.90             NaN 2013-01-05 19:03:24   
15123   19.00             NaN 2014-07-08 15:51:12   
5888    19.30             NaN 2013-08-19 22:56:07   
13297   19.37             NaN 2012-01-03 02:44:15   
2361    19.70             NaN 2012-01-03 02:42:51   
14140   22.00             NaN 2016-05-19 22:08:40   
14381   30.00             NaN 2016-06-25 11:48:24   
13992   32.00             NaN 2012-01-03 02:44:20   
13141   41.00             NaN 2012-01-03 02:44:13   
13723   45.25             NaN 2012-03-02 23:03:06   
2883    81.00             NaN 2017-01-16 18:20:43   
5054   308.00             NaN 2017-05-17 19:56:28   

                                             description  \
4                                                    NaN   
10576  In Britain, seasonal brews for winter are high...   
10537  Named for the witch who guarded the fountain o...   
10382  Our Smoked Wee Heavy is loaded with layers of ...   
1534   Barrel-Aged Imperial Stout aged in Woodford Re...   
1616                                                 NaN   
8901   Latin for "wheat," Triticus is a strong and da...   
8786   Bourbon Barrel-Aged Imperial Last Snow\r\n\r\n...   
8193                                                 NaN   
7757   Imperial stout brewed with chocolate, black tr...   
7750   he Roman Empire had a certain ‘je ne sais quoi...   
774

## IBU

In [22]:
beers.ibu.describe(percentiles=[.025,.25,.5,.75,.975])

count    5902.000000
mean       41.110826
std        26.827465
min         1.000000
2.5%        9.000000
25%        22.000000
50%        33.450000
75%        55.450000
97.5%     100.000000
max       430.000000
Name: ibu, dtype: float64

In [23]:
hist = Histogram(beers[pd.notnull(beers.ibu)].ibu, title = 'IBU Distribution', bins=100,
              x_range=Range1d(0,50))
show(hist)

In [21]:
beers[pd.isnull(beers.description)].count()

abv                          4431
beerVariationId                35
createDate                   6344
description                     0
foodPairings                   22
glasswareId                   792
ibu                          1066
id                           6344
ingredients                    30
isOrganic                    6344
labels                       1689
name                         6344
nameDisplay                  6344
originalGravity               128
servingTemperature            135
servingTemperatureDisplay     135
socialAccounts               1165
srmId                         238
status                       6344
statusDisplay                6344
styleId                      5738
updateDate                   6344
year                           77
createYear                   6344
dtype: int64

In [40]:
### Get Beer Counts by Year Added to Database
count = beers.id.groupby(beers.createYear).count()

In [41]:
count

createYear
2012    4754
2013    1337
2014    2221
2015    3912
2016    1987
2017    1199
Name: id, dtype: int64